### PR TITLE
Boomerang behavior is now defined as a component.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
## About The Pull Request

Fully refactors boomerang behavior to work as a component, so that a thrown boomerang will return back to it's thrower if within range. More than anything this modularizes the behavior for the backend, as the in-game behavior is intentionally the same as before, but shifts the majority of the behavior to the component for neater code.  For unique behavior when being caught or hitting targets, that should remain on the object.

## Why It's Good For The Game

I have a confession to make:
Oztek Boomerangs and the Clown Energy Shield use the exact same code, copied and pasted. I've wanted to move this over to a component for a long time, but wasn't exactly sure how to make that work, honestly. I just put 2 and 2 together on how to get this rolling while I was gone for Christmas and I just got it rolling now.

Plus, now admins have the ability to make anything a boomerang, if they so choose.

## Video

https://user-images.githubusercontent.com/41715314/148668590-fdb6ab8f-de8a-4cd2-b2ef-ce25594e7485.mp4

## Changelog

:cl:
refactor: Boomerang behavior has been moved to a component.
/:cl:

### CODE REVIEW

File: code/datums/components/boomerang.dm [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	. = ..()
+	if(!isitem(parent)) //Only items support being thrown around like a boomerang, feel free to make this apply to humans later on.
+		return COMPONENT_INCOMPATIBLE
+	true_parent = parent
+
+	//Assignments
+	if(_boomerang_throw_range)
+		boomerang_throw_range = _boomerang_throw_range
+	if(_thrower_easy_catch_enabled)
+		thrower_easy_catch_enabled = _thrower_easy_catch_enabled
```
Comment (MrMelbert) [inline code/datums/components/boomerang.dm:27]: - Don't bother saving `true_parent`, just cast it when needed
- underscore parameters are out of style, now parameters are usually just the same as the var name, and we use `src.var = var` to assign them (see the transforming component for an example)

File: code/datums/components/boomerang.dm [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+/datum/component/boomerang/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_MOVABLE_POST_THROW, .proc/PrepareThrow) ///Collect data on current thrower and the throwing datum
+	RegisterSignal(parent, COMSIG_MOVABLE_THROW_LANDED, .proc/ReturnMissedThrow)
+	RegisterSignal(parent, COMSIG_MOVABLE_IMPACT, .proc/ReturnHitThrow)
+	//RegisterSignal(parent, COMSIG_CARBON_SUCCESSFUL_CATCH, .proc/BoomerangCooldown)
```
Comment (MrMelbert) [inline code/datums/components/boomerang.dm:33]: 🤨

File: code/datums/components/boomerang.dm [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+/datum/component/boomerang/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(parent, list(COMSIG_MOVABLE_POST_THROW, COMSIG_MOVABLE_THROW_LANDED, COMSIG_MOVABLE_IMPACT))
+
+/datum/component/boomerang/proc/PrepareThrow(datum/source, atom/thrown_thing, spin)
```
Comment (MrMelbert) [inline code/datums/components/boomerang.dm:39]: - Snake case not camel case please
- Auto-doc your procs for your signals

File: code/datums/components/boomerang.dm [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+/datum/component/boomerang
+	///How far should the boomerang try to travel to return to the thrower?
+	var/boomerang_throw_range = 3
+	///"Thrownthing" datum for the most recent throw.
+	var/datum/thrownthing/thrown_boomerang
+	///If this boomerang is thrown, does it re-enable the throwers throw mode?
+	var/thrower_easy_catch_enabled = FALSE
+	///The object the component's being applied to.
+	var/obj/item/true_parent
+	COOLDOWN_DECLARE(last_boomerang_throw)
```
Comment (MrMelbert) [inline code/datums/components/boomerang.dm:15]: Might as well doc the cooldown as well

File: code/datums/components/boomerang.dm [hunk 5]
```diff
@@ -0,0 +1,76 @@
+#define BOOMERANG_REBOUND_INTERVAL (1 SECONDS)
```
Comment (MrMelbert) [inline code/datums/components/boomerang.dm:1]: Also this define

**Review by MrMelbert — CHANGES_REQUESTED** [review]
This is cool, a few stylistic things I noticed, i'll check out the code itself later

**Comment by Ghommie** [comment]
The cult mirror shield also possesses a similar boomerang behavior. Would it be possible to apply this component to it too?

File: code/datums/components/boomerang.dm [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * If an ojvect is given the boomerang component, it should be thrown back to the thrower after either hitting it's target, or landing on the thrown tile.
+ * Thrown objects should
+ */
+/datum/component/boomerang
+	///How far should the boomerang try to travel to return to the thrower?
+	var/boomerang_throw_range = 3
+	///"Thrownthing" datum for the most recent throw.
+	var/datum/thrownthing/thrown_boomerang
```
Comment (MrMelbert) [inline code/datums/components/boomerang.dm:11]: Can this be a weakref? Seems like it'll be disappearing often

**Review by MrMelbert — CHANGES_REQUESTED** [review]
Got around to looking at it, few concerns and notes

Namely keeping the hard reference to the throwing datum seems like a dangerous idea cause they're likely to perish

Changing the mirror shield to use this would be pretty neat as well

File: code/datums/components/boomerang.dm [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * Proc'd before the thrown is performed in order to gather information regarding each throw as well as handle throw_mode as necessary.
+ * * source: Datum src from original signal call.
+ * * thrown_thing: The atom that has had the boomerang component added to it. Updates thrown_boomerang.
+ * * spin: Carry over from POST_THROW, the speed of rotation on the boomerang when thrown.
+ */
+/datum/component/boomerang/proc/prepare_throw(datum/source, atom/thrown_thing, spin)
+	SIGNAL_HANDLER
+	thrown_boomerang = thrown_thing //Here we update our "thrownthing" datum with that of the original throw for each boomerang. We save it for the return throw.
```
[the remark this answers is not in the document]
Comment (ArcaneMusic) [reply code/datums/components/boomerang.dm:45]: Wow how did this even work the first time 👀

**Comment by ArcaneMusic** [comment]
> The cult mirror shield also possesses a similar(-ish) boomerang behavior. Would it be possible to apply this component to it too?

Looking it it, it appears to use some rather tricky behavior, but it's a bit tricky since he uses more unique system to bounce the throw to new targets that might require more behavior than the current basic boomerang behavior. (Plus they're currently admin only)

File: code/datums/components/boomerang.dm [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+#define BOOMERANG_REBOUND_INTERVAL (1 SECONDS)
+/**
+ * If an ojvect is given the boomerang component, it should be thrown back to the thrower after either hitting it's target, or landing on the thrown tile.
+ * Thrown objects should be thrown back to the original thrower with this component, a number of tiles defined by boomerang_throw_range.
+ */
+/datum/component/boomerang
+	///How far should the boomerang try to travel to return to the thrower?
+	var/boomerang_throw_range = 3
+	///"Thrownthing" datum for the most recent throw.
+	//var/datum/weakref/thrown_boomerang
```
Comment (MrMelbert) [inline code/datums/components/boomerang.dm:11]: 🥴

File: code/datums/components/boomerang.dm [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+	RegisterSignal(parent, COMSIG_MOVABLE_THROW_LANDED, .proc/return_missed_throw)
+	RegisterSignal(parent, COMSIG_MOVABLE_IMPACT, .proc/return_hit_throw)
+
+/datum/component/boomerang/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_MOVABLE_POST_THROW, COMSIG_MOVABLE_THROW_LANDED, COMSIG_MOVABLE_IMPACT))
+
+/**
+ * Proc'd before the thrown is performed in order to gather information regarding each throw as well as handle throw_mode as necessary.
+ * * source: Datum src from original signal call.
+ * * thrown_thing: The atom that has had the boomerang component added to it. Updates thrown_boomerang.
```
Comment (MrMelbert) [inline code/datums/components/boomerang.dm:39]: "The atom"

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/datums/components/boomerang.dm (hunk 8 above)
Comment (ArcaneMusic) [reply code/datums/components/boomerang.dm]: I'll admit, I forgot I committed cringe, my B

File: code/datums/components/boomerang.dm [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+/datum/component/boomerang/proc/return_hit_throw(datum/source, atom/hit_atom, datum/thrownthing/init_throwing_datum)
+	SIGNAL_HANDLER
+	if (!COOLDOWN_FINISHED(src, last_boomerang_throw))
+		return
+	var/obj/item/true_parent = parent
+	var/mob/thrown_by = true_parent.thrownby?.resolve()
+	aerodynamic_swing(init_throwing_datum)
+	if(thrown_by)
+		addtimer(CALLBACK(true_parent, /atom/movable.proc/throw_at, thrown_by, boomerang_throw_range, init_throwing_datum.speed, null, TRUE), 1)
+		COOLDOWN_START(src, last_boomerang_throw, BOOMERANG_REBOUND_INTERVAL)
```
Comment (MrMelbert) [inline code/datums/components/boomerang.dm:63]: Last thing before merging - this code seems to be copy+pasted between `return_hit_throw` and `return_missed_throw`, might be wise to reduce code duplication by moving it into a separate proc?

Also extremely minor nitpick but the `return` at the end isn't necessary

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Pull request merged**